### PR TITLE
IAM: Protect managed service accounts frontend list page

### PR DESCRIPTION
**Pull request by linoman**
**What is this feature?**

This feature disables buttons for managed service accounts at the service accounts list page. Additionally, renames one of the filters from external to managed and aligns the enable/disable buttons at the list page.

**Why do we need this feature?**

With the new managed service accounts, we need to adjust the UI in order to improve the UX.

**Which issue(s) does this PR fix?**:

This PR is part 1 of 2 for the issue https://github.com/grafana/grafana-enterprise/issues/5965

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/serviceaccounts/extsvcaccounts/metrics.go [hunk 1]
```diff
@@ -24,14 +24,14 @@ func newMetrics(reg prometheus.Registerer, saSvc serviceaccounts.Service, logger
 		prometheus.GaugeOpts{
 			Namespace: metricsNamespace,
 			Name:      "extsvc_total",
```
Comment (IevaVasiljeva) [inline pkg/services/serviceaccounts/extsvcaccounts/metrics.go:26]: Nit: should we also rename the metric or do you think it's too late for that?

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx [hunk 2]
```diff
@@ -194,6 +194,7 @@ export const ServiceAccountsListPageUnconnected = ({
             options={[
               { label: 'All', value: ServiceAccountStateFilter.All },
               { label: 'With expired tokens', value: ServiceAccountStateFilter.WithExpiredTokens },
+              { label: 'Managed', value: ServiceAccountStateFilter.Managed },
```
Comment (IevaVasiljeva) [inline public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:197]: Thought: I wonder if it would make sense to filter managed service accounts out of the default service account view? As I imagine users won't interact with them much, so it makes sense to hide them by default. No strong feelings on it though.

**Review by IevaVasiljeva — COMMENTED** [review]
Overall looks good! Left two small suggestions :)

File: pkg/services/serviceaccounts/extsvcaccounts/metrics.go (hunk 1 above)
Comment (gamab) [reply pkg/services/serviceaccounts/extsvcaccounts/metrics.go:26]: I'd rather keep the name as it is. As of today external service accounts are the only managed service accounts, but that might not be true tomorrow. Or am I overthinking this?

File: pkg/services/serviceaccounts/database/store.go [hunk 3]
```diff
@@ -318,7 +318,7 @@ func (s *ServiceAccountsStoreImpl) SearchOrgServiceAccounts(ctx context.Context,
 				whereConditions,
 				"is_disabled = ?")
 			whereParams = append(whereParams, s.sqlStore.GetDialect().BooleanStr(true))
-		case serviceaccounts.FilterOnlyExternal:
+		case serviceaccounts.FilterOnlyManaged:
```
Comment (gamab) [inline pkg/services/serviceaccounts/database/store.go:321]: Why did you rename the filter?
I was thinking we should consider external and managed to be two different concepts. 
As of today external service accounts are the only managed service accounts, but that might not be true tomorrow.

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx (hunk 2 above)
Comment (gamab) [reply public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:197]: I like the idea. Maybe we can put a pin on it and tackle it in another PR? (As I feel it'd add another filter)

Another thing to consider, we are probably talking about 10 external service accounts top (if you have 10 plugins installed that need one :D)

File: pkg/services/serviceaccounts/extsvcaccounts/metrics.go (hunk 1 above)
Comment (IevaVasiljeva) [reply pkg/services/serviceaccounts/extsvcaccounts/metrics.go:26]: Ah, interesting, I didn't realise that we might have other types of external SAs in the future. Do we have any concrete plans?

Comment (gamab) [reply pkg/services/serviceaccounts/extsvcaccounts/metrics.go:26]: > Do we have any concrete plans

No :smile: hence the:

> Or am I overthinking this?

I was thinking of this like roles, at the beginning we only had `fixed` roles, but then came `managed`, `extsvc`, `plugins`... All of these are technically managed by Grafana, and this is all the frontend should care about. Yet they have different naming patterns in the backend.

Comment (linoman) [reply pkg/services/serviceaccounts/extsvcaccounts/metrics.go:26]: After some conversations around the topic we will not rename anything `managed` but instead `external`. We feel we're thinking way ahead of time.

Comment (linoman) [reply pkg/services/serviceaccounts/extsvcaccounts/metrics.go:26]: I'll be undoing some of the renames

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx (hunk 2 above)
Comment (linoman) [reply public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:197]: I like this idea. I agree with @gamab, let's tackle this in another PR.

My thinking is, we need to be more transparent with the users. I don't think the filters tell us there are any external SA until we click on it. I think it'd be really cool for the filters to append the total count of the SA queried.

File: public/app/features/serviceaccounts/components/ServiceAccountsListItem.tsx [hunk 4]
```diff
@@ -112,32 +112,48 @@ const ServiceAccountListItem = memo(
...
+            <HorizontalGroup justify="flex-end">
               <IconButton
-                className={styles.deleteButton}
-                name="trash-alt"
+                disabled={true}
+                name="lock"
                 size="md"
-                onClick={() => onRemoveButtonClick(serviceAccount)}
-                tooltip={`Delete service account ${serviceAccount.name}`}
+                tooltip={`This is a managed Service account and cannot be modified.`}
```
Comment (gamab) [inline public/app/features/serviceaccounts/components/ServiceAccountsListItem.tsx:153]:
```suggestion
                tooltip={`This is a managed service account and cannot be modified.`}
```

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx [hunk 5]
```diff
@@ -194,6 +194,7 @@ export const ServiceAccountsListPageUnconnected = ({
             options={[
               { label: 'All', value: ServiceAccountStateFilter.All },
               { label: 'With expired tokens', value: ServiceAccountStateFilter.WithExpiredTokens },
+              { label: 'Managed', value: ServiceAccountStateFilter.External },
```
Comment (gamab) [inline public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:197]: We should add this behind a feature check I think :slightly_smiling_face:

File: pkg/services/serviceaccounts/api/api.go [hunk 6]
```diff
@@ -265,10 +265,14 @@ func (api *ServiceAccountsAPI) SearchOrgServiceAccountsWithPaging(c *contextmode
 	// its okay that it fails, it is only filtering that might be weird, but to safe quard against any weird incoming query param
 	onlyWithExpiredTokens := c.QueryBool("expiredTokens")
 	onlyDisabled := c.QueryBool("disabled")
+	onlyExternal := c.QueryBool("external")
 	filter := serviceaccounts.FilterIncludeAll
 	if onlyWithExpiredTokens {
 		filter = serviceaccounts.FilterOnlyExpiredTokens
 	}
+	if onlyExternal {
```
Comment (gamab) [inline pkg/services/serviceaccounts/api/api.go:273]: If we add a feature check here, that'll leave us room for change.

**Review by gamab — COMMENTED** [review]
LGTM! I think we just miss one or two feature checks and we are good to go.

File: public/api-enterprise-spec.json [hunk 7]
```diff
@@ -4294,7 +4294,7 @@
           "type": "integer",
           "format": "int64"
         },
-        "isManaged": {
+        "isExternal": {
```
Comment (gamab) [inline public/api-enterprise-spec.json:4297]:
```suggestion
        "isManaged": {
```

Comment (linoman) [reply public/api-enterprise-spec.json]: 🤦🏻 fixed!

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx [hunk 8]
```diff
@@ -56,6 +57,16 @@ const mapDispatchToProps = {
 
 const connector = connect(mapStateToProps, mapDispatchToProps);
 
+const availableFilters = [
+  { label: 'All', value: ServiceAccountStateFilter.All },
+  { label: 'With expired tokens', value: ServiceAccountStateFilter.WithExpiredTokens },
+  { label: 'Disabled', value: ServiceAccountStateFilter.Disabled },
+];
+
+if (config.featureToggles.externalServiceAccounts) {
```
Comment (gamab) [inline public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:66]:
```suggestion
if (config.featureToggles.externalServiceAccounts || config.featureToggles.externalServiceAuth) {
```

File: public/openapi3.json [hunk 9]
```diff
@@ -4791,7 +4791,7 @@
             "format": "int64",
             "type": "integer"
           },
-          "isManaged": {
+          "isExternal": {
```
Comment (gamab) [inline public/openapi3.json:4794]:
```suggestion
          "isManaged": {
```

File: pkg/services/serviceaccounts/api/api.go (hunk 6 above)
Comment (linoman) [reply pkg/services/serviceaccounts/api/api.go]: done

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx (hunk 5 above)
Comment (linoman) [reply public/app/features/serviceaccounts/ServiceAccountsListPage.tsx]: done

File: public/app/features/serviceaccounts/components/ServiceAccountsListItem.tsx (hunk 4 above)
Comment (linoman) [reply public/app/features/serviceaccounts/components/ServiceAccountsListItem.tsx]: done

**Review by gamab — APPROVED** [review]
Great work lino!
